### ISSUE TITLE
Integrating support for graphene-sqlalchemy-filter into code base

### ISSUE THREAD

**Issue by palisadoes** [graphql-python/graphene-sqlalchemy#335]
As there is limited filtering with [graphene-sqlalchemy](https://github.com/graphql-python/graphene-sqlalchemy), I have been using [graphene-sqlalchemy-filter](https://github.com/art1415926535/graphene-sqlalchemy-filter) which has many stock filters included. Unfortunately the developer seems to have lost interest in the code base and a critical PR to add support for Graphene 3.0 has been unmerged for six months. 

- https://github.com/art1415926535/graphene-sqlalchemy-filter/pull/63

I have tried contacting the developer a few times without response.

As the filtering code for the add-on is contained in the `graphene-sqlalchemy-filter` package's two files, I propose merging this code into `graphene-sqlalchemy`.

Would this be OK?

**Comment by erikwrede** [comment]
I just followed your link from the filter repo, so I thought I might leave a reply here as well :)

While in general I support the idea of adding filters to graphene-sqa, and like the approach that the author has taken, I think a simple merge won't do it. The filters should be deeply integrated into this repo to allow for even more features, so this has to be planned. But I think using this library as inspiration would be a great start. 
Maybe we can work something out together? Or maybe one of the maintainers of this repo has any supporting or opposing ideas to consider before starting. It would be sad to see yet another merge request get stuck unapproved for a long time.

**Comment by palisadoes** [comment]
Thanks. My python skills are good, but not great. However, I'd be willing to learn. @cooldudemcgeexl would you be interested in a collaboration?

**Comment by palisadoes** [comment]
We have an interested party in adding the filtering syntax.

- https://github.com/graphql-python/graphene-sqlalchemy/pull/164#issuecomment-1083510163

**Comment by erikwrede** [comment]
Sounds great!

I think the current external library is great but still has some places where the design lacks flexibility.
Maybe we should find some interested users and collect their filtering requirements here as a start?

**Comment by erikwrede** [comment]
Implementation is in progress. The main PR will be #357. Please join us on Slack `#graphene-sqlalchemy-filter` if you would like to contribute or have anything to add!

**Comment by fred-asimov** [comment]
Just bumping this to see where things are. I'm also using a fork of `graphene-sqlalchemy-filter` with some patches, including a small fix for the conversion refactor, but would love to see filters fully integrating and working. Happy to pitch in if needed!

**Comment by palisadoes** [comment]
@polgfred Coincidentally we had a meeting yesterday where we decided it's time to start alpha testing this integration. 

1. Use the add-filters branch: https://github.com/graphql-python/graphene-sqlalchemy/tree/add-filters
2. We don't have documentation yet, but should have it soon. You'll need to use the test cases as the interim documentation: https://github.com/graphql-python/graphene-sqlalchemy/blob/add-filters/graphene_sqlalchemy/tests/test_filters.py
3. You'll notice that the original filter uses the `filters` keyword, we are using the `filter` keyword. The rationale for this is so that both methodologies can work side by side as people transition (assuming the package dependencies are frozen)

@erikwrede, @sabard  can you provide the link to the original syntax guide we are using? That will help greatly until we get the documentation updated.

**Comment by erikwrede** [comment]
@polgfred cadu should have the syntax guide via google docs. If you want access too, please send me your email address via Slack 🙂

**Comment by palisadoes** [comment]
@polgfred These are the syntax guides we used as the templates for the work
- https://www.gatsbyjs.com/docs/query-filters/
- https://dgraph.io/docs/graphql/queries/and-or-not/#nesting

**Comment by adiberk** [comment]
I have been testing filtering on the latest version of graphene-sqlalchemy (So excited to migrate to v3!)
However I noticed that I am unable to perform query joins in the new filter setup.  In the  setup with graphene-sqlalchemy-filters. You were allowed you to add join statements to a query in the filter.

For example
```    def assigned_to_borrower_filter(
        info: LevResolveInfo, query: Query, assigned_to_borrower_filter: bool
    ) -> tuple[Query, Optional[ColumnOperators]]:
        _filter = None
        if assigned_to_borrower_filter:
            query = query.join(TaskDealUser).join(DealUser).join(User)
            _filter = User.role == UserRole.borrower
        return query, _filter
```


Would it be valuable to allow returning a tuple of the filter and the query itself at each loop through the filters?  And we make returning either one optional (ie you could just return the query or just the filter or both).
Maybe my implementation doesn't make much sense but we have used it like this a couple of times in our codebase
